upgrade/v12xto130/upgrade.go

```diff
@@ -0,0 +1,72 @@
+package v12xto130
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	upgradeLogPrefix = "upgrade from v1.2.x to v1.3.0: "
+)
+
+func UpgradeResources(namespace string, lhClient *lhclientset.Clientset, kubeClient *clientset.Clientset) error {
+	if err := upgradeReplicas(namespace, lhClient); err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeReplicas(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"upgrade replica failed")
+	}()
+
+	expireInSecond := 60
+	endTime := time.Now().Add(time.Duration(expireInSecond) * time.Second)
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	for time.Now().Before(endTime) {
+		<-ticker.C
+		replicas, err := lhClient.LonghornV1beta2().Replicas(namespace).List(context.TODO(), metav1.ListOptions{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil
+			}
+			return errors.Wrapf(err, "failed to list all existing Longhorn replicas during the replica upgrade")
+		}
+
+		retry := false
+		for _, r := range replicas.Items {
+			if r.Status.IP == "" {
+				continue
+			}
+
+			if r.Status.StorageIP != "" {
+				continue
+			}
+
+			r.Status.StorageIP = r.Status.IP
+			_, err = lhClient.LonghornV1beta2().Replicas(namespace).Update(context.TODO(), &r, metav1.UpdateOptions{})
+			if err != nil {
+				retry = true
```

Comment:
Thanks for the prompt update!

However, I am thinking probably we should make it simply failed, and then users can wait for the next manager pod to restart to retry the upgrade because the upgrade should be idempotent instead of using a specific retry in this upgradeReplica function. 

WDYT?

Reply:
We can do that but I am not sure if it is worth going through the Kubernetes pod lifecycle when caused by conflicting errors. Maybe we can still retry because this will expire quickly(1m) and return an error when failed to update?

Reply:
Updated.